src/sqlancer/StatementReducer.java

```diff
@@ -28,32 +33,59 @@ public void reduce(G state, Reproducer<G> reproducer, G newGlobalState) throws E
         printQueries(knownToReproduceBugStatements);
         System.out.println();
 
-        do {
+        if (knownToReproduceBugStatements.size() <= 1) {
+            return;
+        }
+
+        partitionNum = 2;
+        while (knownToReproduceBugStatements.size() >= 2) {
             observedChange = false;
+
             knownToReproduceBugStatements = tryReduction(state, reproducer, newGlobalState,
-                    knownToReproduceBugStatements, (candidateStatements, i) -> {
-                        candidateStatements.remove((int) i);
+                    knownToReproduceBugStatements, (candidateStatements, s, l) -> {
+                        List<Query<C>> removed = new ArrayList<>();
+                        for (int i = 0; i < candidateStatements.size(); i++) {
+                            if (i >= s + l || i < s) {
+                                removed.add(candidateStatements.get(i));
+                            }
+                        }
+                        candidateStatements.clear();
+                        candidateStatements.addAll(removed);
                         return true;
                     });
-        } while (observedChange);
+            if (!observedChange) {
+                if (partitionNum == knownToReproduceBugStatements.size()) {
+                    break;
+                }
+                // increase the search granularity
+                partitionNum = Math.min(partitionNum * 2, knownToReproduceBugStatements.size());
+            }
+        }
 
         System.out.println("Reduced query:");
         printQueries(knownToReproduceBugStatements);
     }
 
     private List<Query<C>> tryReduction(G state, // NOPMD
             Reproducer<G> reproducer, G newGlobalState, List<Query<C>> knownToReproduceBugStatements,
-            BiFunction<List<Query<C>>, Integer, Boolean> reductionOperation) throws Exception {
+            TriFunction<List<Query<C>>, Integer, Integer, Boolean> reductionOperation) throws Exception {
 
         List<Query<C>> statements = knownToReproduceBugStatements;
-        for (int i = 0; i < statements.size(); i++) {
+
+        int start = 0;
+        int subLength = statements.size() / partitionNum;
```

Comment:
The variable `subLength` is outside of the loop. Doesn't it need to change when the loop goes on?

Reply:
Delta-debugging split the hole array into several blocks and try remove one of them to see if the same error is still here. The block could be large at the begining(e.g. half of the array size) if each block removal let the bug disappear, shrink the block size.

```subLength``` is the block size. It determines how many statements should be dropped. So it should be fixed in the inner loop.